sbi/neural_nets/embedding_nets.py

```diff
@@ -29,5 +39,168 @@ def forward(self, x: Tensor) -> Tensor:
+
+        # construct fully connected layers
+        fc_layers = [
+            nn.Linear(
+                out_channels_cnn_2 * (int(input_dim / out_channels_cnn_2)), num_hiddens
+            ),
+            nn.ReLU(),
+        ]
+
+        # first and last layer is defined by the input and output dimension.
+        # therefor the "number of hidden layeres" is num_layers-2
+        for _ in range(num_fully_connected - 2):
+            fc_layers.append(nn.Linear(num_hiddens, num_hiddens))
+            fc_layers.append(nn.ReLU())
+        fc_layers.append(nn.Linear(num_hiddens, output_dim))
+
+        self.fc_subnet = nn.Sequential(*fc_layers)
+
+    def forward(self, x):
+        """Network forward pass.
+        Args:
+            x: Input tensor (batch_size, input_dim)
+        Returns:
+            Network output (batch_size, output_dim).
+        """
+        x = self.conv_subnet(x)
+        x = torch.flatten(x, 1)  # flatten all dimensions except batch
+        embedding = self.fc_subnet(x)
+
+        return embedding
+
+
+class PermutationInvariantEmbedding(nn.Module):
+    """Permutation invariant embedding network.
+
+    Takes as input a tensor with (batch, permutation_dim, input_dim)
+    and outputs (batch, output_dim).
+    """
+
+    def __init__(
+        self,
+        single_trial_net: nn.Module,
+        input_dim: int,
+        output_dim: int,
+        num_fully_connected: int = 2,
+        num_hiddens: int = 20,
+        combining_operation: str = "mean",
+    ):
+        """Permutation invariant multi-layer NN.
+            The single_trial_net is applied to each "trial" of the input
+            and is combined by the combining_operation (mean or sum).
+
+        Args:
+            single_trial_net: Network to process one trial, the combining_operation is
+                applied to its output. Taskes as input (batch, input_dim).
+                Remark: This network should be large enough as it acts on all (iid)
+                inputs seperatley and needs enough capacity to process the information
+                of all inputs.
+            input_dim: Dimensionality of input to the fully connected layers
+                (output_dimension of single_trial_net).
+            output_dim: Dimensionality of the output.
+            num_fully_connected: Number of fully connected layer, minimum of 2.
+            num_hiddens: Number of hidden dimensions in fully-connected layers.
+            combining_operation: How to combine the permutational dimensions, one of
+                'mean' or 'sum'.
+        """
+        super().__init__()
+        self.single_trial_net = single_trial_net
+        self.input_dim = input_dim
+        self.output_dim = output_dim
+        self.num_hiddens = num_hiddens
+        self.combining_operation = combining_operation
+
+        if combining_operation not in ["sum", "mean"]:
+            raise ValueError("Please enter a valdi combining operation.")
```

Comment:
typo: valid

Reply:
to make the error more explicit, maybe: `combining_operation must be in ['sum', 'mean']`